fortanix-vme/fortanix-vme-runner/src/lib.rs

```diff
@@ -89,14 +89,98 @@ impl Listener {
     }
 }
 
+#[derive(Debug)]
+struct Connection {
+    tcp_stream: TcpStream,
+    vsock_stream: VsockStream<Std>,
+    remote_name: String,
+}
+
+impl Connection {
+    fn new(vsock_stream: VsockStream<Std>, tcp_stream: TcpStream, remote_name: String) -> Self {
+        Connection {
+            tcp_stream,
+            vsock_stream,
+            remote_name,
+        }
+    }
+
+    fn close(&self) {
+        let _ = self.tcp_stream.shutdown(Shutdown::Both);
+        let _ = self.vsock_stream.shutdown(Shutdown::Both);
+    }
+
+    /// Exchanges messages between the remote server and enclave. Returns `true` when the
+    /// connection should remain active, false otherwise
+    fn proxy(&mut self) -> bool {
+        fn exchange<S: StreamConnection, D: StreamConnection>(src: &mut S, src_name: &str, dst: &mut D, dst_name: &str) -> bool {
+            // According to the `Read` threat documentation, reading 0 bytes
+            // indicates that the connection has been shutdown correctly. So we
+            // close the proxy service
+            // https://doc.rust-lang.org/std/io/trait.Read.html#tymethod.read
+            match Server::transfer_data(src, src_name, dst, dst_name) {
+                Ok(n) if n == 0 => false,
+                Ok(_)           => true,
+                Err(_)          => false,
+            }
+        }
+        let remote = self.tcp_stream.as_raw_fd();
+        let proxy = self.vsock_stream.as_raw_fd();
+
+        let mut read_set = FdSet::new();
+        read_set.insert(remote);
+        read_set.insert(proxy);
+
+        if let Ok(_num) = select(None, Some(&mut read_set), None, None, None) {
+            if read_set.contains(remote) {
+                return exchange(&mut self.tcp_stream, &self.remote_name, &mut self.vsock_stream, "proxy");
+            }
+            if read_set.contains(proxy) {
+                return exchange(&mut self.vsock_stream, "proxy", &mut self.tcp_stream, &self.remote_name);
+            }
```

Comment:
Is there a risk for starvation of one side of the connection here? Suppose that the `read_set` continually contains `remote` each time `fn proxy()` is called, then we'd never get to exchange data in the other direction. Probably that's okay according to the TCP spec but it might be nicer to still do the exchange from the vsock stream to the TCP stream even if there was an exchange in the other direction.

Reply:
Yes that could be a problem. Good catch!